model/flow/service_event.go

```diff
@@ -61,69 +69,87 @@ func (sel ServiceEventList) EqualTo(other ServiceEventList) (bool, error) {
 	return true, nil
 }
 
+// ServiceEventMarshaller marshals and unmarshals all types of service events.
 type ServiceEventMarshaller interface {
-	Unmarshal(b []byte) (ServiceEvent, error)
-	UnmarshalWithType(
-		b []byte,
-		eventType ServiceEventType,
-	) (
-		ServiceEvent,
-		error,
-	)
+	// UnmarshalWrapped unmarshals the service event and returns it as a wrapped ServiceEvent type.
+	// The input bytes must be encoded as a generic wrapped ServiceEvent type.
+	UnmarshalWrapped(b []byte) (ServiceEvent, error)
+	// UnmarshalWithType unmarshals the service event and returns it as a wrapped ServiceEvent type.
+	// The input bytes must be encoded as a specific event type (for example, EpochSetup).
+	UnmarshalWithType(b []byte, eventType ServiceEventType) (ServiceEvent, error)
 }
 
 type marshallerImpl struct {
-	MarshalFunc   func(v interface{}) ([]byte, error)
-	UnmarshalFunc func(data []byte, v interface{}) error
+	marshalFunc   func(v interface{}) ([]byte, error)
+	unmarshalFunc func(data []byte, v interface{}) error
 }
 
 var (
 	ServiceEventJSONMarshaller = marshallerImpl{
-		MarshalFunc:   json.Marshal,
-		UnmarshalFunc: json.Unmarshal,
+		marshalFunc:   json.Marshal,
+		unmarshalFunc: json.Unmarshal,
 	}
 	ServiceEventMSGPACKMarshaller = marshallerImpl{
-		MarshalFunc:   msgpack.Marshal,
-		UnmarshalFunc: msgpack.Unmarshal,
+		marshalFunc:   msgpack.Marshal,
+		unmarshalFunc: msgpack.Unmarshal,
 	}
 	ServiceEventCBORMarshaller = marshallerImpl{
```

Comment:
could we add a comment here that Json and MsgPack are to be used only for trusted data sources. Due to the BFT nature of the protocol, I we should use RLP for protocol-internal encodings (which I think is the case). Not sure whether CBOR is BFT or not (@fxamacker would know).

Reply:
I will add a comment. Depends on what you mean by protocol-internal, but I don't think we do use RLP in all these cases. or example, we use CBOR (not RLP) for encoding messages sent over the network. 

For the KV Store serialization issue, I looked into where we were using different serialization methods, it's summarized here: https://github.com/onflow/flow-go/issues/5305#issuecomment-1917635711.

Reply:
> could we add a comment here that Json and MsgPack are to be used only for trusted data sources. Due to the BFT nature of the protocol, I we should use RLP for protocol-internal encodings (which I think is the case). Not sure whether CBOR is BFT or not (@fxamacker would know).

Hey @AlexHentschel here's a bit about CBOR security, deterministic encoding requirements, uses, etc.  I defer to you about BFT.

- CBOR is used in [Shelley Networking Protocol (v1.3.0 Jan 2024)](https://ouroboros-network.cardano.intersectmbo.org/pdfs/network-spec/network-spec.pdf) which is newer than [Ourborus-BFT](https://eprint.iacr.org/2018/1049.pdf)
  > All mini-protocols are encoded using concise binary object representation (CBOR)

- CBOR is also used in Passkeys, WebAuthn, COSE ([IETF STD96](https://www.rfc-editor.org/info/std96)), and other security-related standards.

Flow is using CBOR for some networking and storage (onflow/atree, onflow/cadence, onflow/ccf, onflow/flow-go). The CBOR codec we use has configurable settings for security, deterministic encoding, etc.

<details><summary>JSON → MSGPACK → CBOR</summary>

## JSON → MSGPACK → CBOR

CBOR was inspired by JSON and MsgPack so it benefits from hindsight.

CBOR is an extensively reviewed IETF Internet Standard designed with security considerations in mind:
- RFC7049 (Oct 2013) -> 7049bis -> [IETF STD94 / RFC8949](https://www.rfc-editor.org/info/std94) (Dec 2020).

Unlike JSON and MsgPack, CBOR defines [core deterministic encoding requirements](https://www.rfc-editor.org/rfc/rfc8949.html#name-core-deterministic-encoding).  This allows CBOR encoders written in different programming languages to produce identical deterministic encodings.

## Efficient Storage

[IETF RFC8618](https://datatracker.ietf.org/doc/html/rfc8618) chose CBOR after [evaluating binary data formats](https://datatracker.ietf.org/doc/html/rfc8618#appendix-C) for efficient storage of messages.

> The format is designed for efficient storage and transmission of large packet captures of DNS traffic; it attempts to minimize the size of such packet capture files but retain the full DNS message contents along with the most useful transport metadata.

</details>

Reply:
Thanks @fxamacker for the context